Add experimental ggml-hexagon backend for the Hexagon NPU

max-krasnyansky opened the pull request:
This PR introduces a new experimental backend `ggml-hexagon` with support for the Hexagon NPU.

Highlights:
- Supports Hexagon versions: v73, v75, v79, and v81
- Targets Android devices based on Snapdragon SoCs: Gen3, 8-Elite, and 8-Elite Gen5
- Supports Q4_0, Q8_0, MXFP4, and FP32 data types
- Implements core LLM ops: MUL_MAT/MUL_MAT_ID, ADD/SUB/MUL/ADD_ID, RMS_NORM, ROPE, GLU/SWIGLU, SOFTMAX
- Minimal build dependencies (just needs Android NDK and Hexagon-SDK Community Edition) 

**Note:** This backend is experimental and may exhibit instability or limited performance across supported devices.
It is intended for early testing and feedback from llama.cpp/ggml developer and user community.

Please see (docs/backend/hexagon/README.md) for build and basic usage info.
Also (docs/backend/hexagon/developer.md) for some notes on things like buffer management, etc.

Tested with the following models:

- Llama-3.2-1B-Instruct-Q4_0.gguf
- Llama-3.2-3B-Instruct-Q4_0.gguf
- Llama-3.1-8B-Instruct-Q4_0.gguf
- Qwen3-4B-Q4_0.gguf
- Qwen3-8B-128K-Q4_0.gguf
- Qwen3-14B-128K-Q4_0.gguf
- LFM2-1.2B-Q4_0.gguf
- OLMoE-1B-7B-0125-Instruct-Q4_0.gguf
- gpt-oss-20b-mxfp4.gguf & gpt-oss-20b.mxfp4-q4_0.gguf  (requires latest devices with 16+ GB of DDR)

Known-issues:

- Test-backend-ops failures for supported Ops. There are a few corner-cases that we don't handle yet which doesn't affect the models listed above. We will follow up with fixes for that.
- Tensor-override option needs some updates to work with the `HTP-REPACK` buffers (see some notes/questions below)
- Integration (buffer sharing, etc) with the OpenCL/Adreno backend needs work

Future-work:

- More optimizations (kernels, fusion, etc), more Ops (SHORTCONV, etc), more DataTypes
- Better integration with the OpenCL/Adreno backend
- Support for Windows on Snapdragon devices 


---
@slaren 
It'd be good to make `extra_buffers_type` a bit better exposed/integrated. I started thinking that `device_get_buffer_type` should probably just return a list of buffer types. Currently only the CPU backend needed those extra bufs but now the Hexagon needs them too (ie all buffers are actually normal host buffers from CPU perspective and extras are needed only to force the REPACK). I added basic support for that in the model loader (separate commit here in the PR) and was going to sprinkle some more but we should discuss if there is perhaps a better way to handle this.

I included some wrapper scripts (under docs/backends/hexagon) to make running stuff over ADB easier. Let me know if we should put them in some other directory. They are a bit Snapdragon specific because of the env vars needed to find Hexagon/HTP libraries (described in the developer.md). 

---
@ggerganov 
I included a commit that groups the Attention and FFN MATMUL Ops. 
Like this:
```
node #841 ( MUL_MAT):  Qcur-27 (   1M) [ HTP0 ] use=1: blk.27.attn_q.weight [ HTP0 ]  attn_norm-27 [ HTP0 ] 
node #842 ( MUL_MAT):  Kcur-27 ( 512K) [ HTP0 ] use=1: blk.27.attn_k.weight [ HTP0 ]  attn_norm-27 [ HTP0 ] 
node #843 ( MUL_MAT):  Vcur-27 ( 512K) [ HTP0 ] use=1: blk.27.attn_v.weight [ HTP0 ]  attn_norm-27 [ HTP0 ] 
node #845 (    ROPE):  Qcur-27 (   1M) [ HTP0 ] use=1:   Qcur-27 (reshaped) [ HTP0 ] HTP0#leaf_6#0 [ NULL ] 
```
This allows us to easily reuse dynamically quantized `attn_norm-27`. Hexagon/HTP backend places quantized tensors in VTCM (basically a SW managed cache) where subsequent Ops can reuse it.
This change doesn't seem to have any affect on other backends (I did some basic checks with CPU, OpenCL, CUDA, and Metal). Please let me know what you think. Perhaps, you have suggestions for how to do this better.

---
Marking as a Draft for now because I'm working on enabling the CI. 
Otherwise all required bits and pieces are ready to go.

---
<details>

<summary>Some outputs of the Hexagon Backend in action on my Galaxy S25+</summary>

```
~/src/llama.cpp-hexagon$ M=../gguf/Llama-3.2-1B-Instruct-Q4_0.gguf D=HTP0 docs/backend/hexagon/run-cli.sh -no-cnv -p \"what is the most popular cookie in the world?\"
+ adb shell ...
ggml_opencl: selected platform: 'QUALCOMM Snapdragon(TM)'
ggml_opencl: device: 'QUALCOMM Adreno(TM) 830 (OpenCL 3.0 Adreno(TM) 830)'
...
ggml-hex: allocating new registry : ndev 1
ggml-hex: HTP arch version v79
ggml-hex: allocating new session: HTP0
ggml-hex: new session: HTP0 : session-id 0 domain-id 3 uri file:///libggml-htp-v79.so?htp_iface_skel_handle_invoke&_modver=1.0&_dom=cdsp&_session=0 handle 0xb400007974ffcd90
build: 6733 (6a8cf8914) with Android (13324770, +pgo, +bolt, +lto, +mlgo, based on r530567d) clang version 19.0.0 (https://android.googlesource.com/toolchain/llvm-project 97a699bf4812a18fb657c2779f5296a4ab2694d2) for x86_64-unknown-linux-gnu
...
load_tensors: offloaded 17/17 layers to GPU
load_tensors:          CPU model buffer size =   225.49 MiB
load_tensors:         HTP0 model buffer size =     0.26 MiB
load_tensors:  HTP0-REPACK model buffer size =   504.00 MiB
...
llama_context: n_ctx_per_seq (8192) < n_ctx_train (131072) -- the full capacity of the model will not be utilized
llama_context:        CPU  output buffer size =     0.49 MiB
llama_kv_cache:       HTP0 KV buffer size =   136.00 MiB
llama_kv_cache: size =  136.00 MiB (  8192 cells,  16 layers,  1/1 seqs), K (q8_0):   68.00 MiB, V (q8_0):   68.00 MiB
llama_context:       HTP0 compute buffer size =    15.00 MiB
llama_context:        CPU compute buffer size =    62.62 MiB
llama_context: graph nodes  = 503
llama_context: graph splits = 41
...
 Chocolate chip cookies are the most popular cookie in the world. According to the International Association of Culinary Professionals, chocolate chip cookies are the most popular cookie in the world.
In the United States, chocolate chip cookies are a beloved favorite, and many bakeries and restaurants offer them as a classic treat. They are often associated with family gatherings, road trips, and comfort food.
In fact, according to a survey conducted by YouGov in 2020, chocolate chip cookies are the most popular cookie in the United States, with 27% of respondents naming them as their favorite. This makes chocolate chip cookies the top choice among Americans, and a close second in other countries. [end of text]


llama_perf_sampler_print:    sampling time =       7.12 ms /   147 runs   (    0.05 ms per token, 20634.48 tokens per second)
llama_perf_context_print:        load time =     623.27 ms
llama_perf_context_print: prompt eval time =      81.60 ms /    11 tokens (    7.42 ms per token,   134.81 tokens per second)
llama_perf_context_print:        eval time =    2428.18 ms /   135 runs   (   17.99 ms per token,    55.60 tokens per second)
llama_perf_context_print:       total time =    2630.13 ms /   146 tokens
llama_perf_context_print:    graphs reused =        134
llama_memory_breakdown_print: | memory breakdown [MiB] | total   free    self   model   context   compute    unaccounted |
llama_memory_breakdown_print: |   - HTP0 (Hexagon)     |  2048 = 2048 + (   0 =     0 +       0 +       0) +           0 |
llama_memory_breakdown_print: |   - Host               |                  439 =   225 +     136 +      77                |
llama_memory_breakdown_print: |   - HTP0-REPACK        |                  504 =   504 +       0 +       0                |



~/src/llama.cpp-hexagon$ M=LFM2-1.2B-Q4_0.gguf D=HTP0 docs/backend/hexagon/run-cli.sh -no-cnv -p \"what is the most popular cookie in the world?\"
+ adb shell ...
...
load_tensors: offloaded 17/17 layers to GPU
load_tensors:          CPU model buffer size =   105.24 MiB
load_tensors:         HTP0 model buffer size =     0.25 MiB
load_tensors:  HTP0-REPACK model buffer size =   555.75 MiB
...
llama_context: n_ctx_per_seq (8192) < n_ctx_train (128000) -- the full capacity of the model will not be utilized
llama_context:        CPU  output buffer size =     0.25 MiB
llama_kv_cache:       HTP0 KV buffer size =    51.00 MiB
llama_kv_cache: size =   51.00 MiB (  8192 cells,   6 layers,  1/1 seqs), K (q8_0):   25.50 MiB, V (q8_0):   25.50 MiB
llama_memory_recurrent:       HTP0 RS buffer size =     0.16 MiB
llama_memory_recurrent: size =    0.16 MiB (     1 cells,  16 layers,  1 seqs), R (f32):    0.16 MiB, S (f32):    0.00 MiB
llama_context:       HTP0 compute buffer size =    14.00 MiB
llama_context:        CPU compute buffer size =    32.00 MiB
llama_context: graph nodes  = 549
llama_context: graph splits = 49
...
system_info: n_threads = 4 (n_threads_batch = 4) / 8 | CPU : NEON = 1 | ARM_FMA = 1 | MATMUL_INT8 = 1 | DOTPROD = 1 | REPACK = 1 | 

...
**(a) Chocolate chip** is the answer, but keep in mind that preferences can vary widely by region and individual taste.

For the most accurate and current data, you might want to look into recent market analyses or consumer surveys. However, chocolate chip remains a strong contender due to its universal appeal and cultural significance. [end of text]


llama_perf_sampler_print:    sampling time =       9.49 ms /   301 runs   (    0.03 ms per token, 31720.94 tokens per second)
llama_perf_context_print:        load time =     694.86 ms
llama_perf_context_print: prompt eval time =      76.28 ms /    11 tokens (    6.93 ms per token,   144.20 tokens per second)
llama_perf_context_print:        eval time =    4698.82 ms /   289 runs   (   16.26 ms per token,    61.50 tokens per second)
llama_perf_context_print:       total time =    4793.75 ms /   300 tokens
llama_perf_context_print:    graphs reused =          0
llama_memory_breakdown_print: | memory breakdown [MiB] | total   free    self   model   context   compute    unaccounted |
llama_memory_breakdown_print: |   - HTP0 (Hexagon)     |  2048 = 2048 + (   0 =     0 +       0 +       0) +           0 |
llama_memory_breakdown_print: |   - Host               |                  202 =   105 +      51 +      46                |
llama_memory_breakdown_print: |   - HTP0-REPACK        |                  555 =   555 +       0 +       0                |
```

</details>

max-krasnyansky committed:
hexagon: add experimental ggml-hexagon backend for the Hexagon NPU

This commit introduces a new experimental backend `ggml-hexagon` with support for the Hexagon NPU.

Highlights:
- Supports Hexagon versions: v73, v75, v79, and v81
- Targets Android devices based on Snapdragon SoCs: Gen3, 8-Elite, and 8-Elite Gen5
- Supports Q4_0, Q8_0, MXFP4, and FP32 data types
- Implements core LLM ops: MUL_MAT/MUL_MAT_ID, ADD/SUB/MUL/ADD_ID, RMS_NORM, ROPE, GLU/SWIGLU, SOFTMAX

**Note:** This backend is experimental and may exhibit instability or limited performance across supported devices.
It is intended for early testing and feedback from llama.cpp/ggml developer and user community.

Co-Authored-By: Rajdeep Ganguly <rganguly@qti.qualcomm.com>
Co-Authored-By: Todor Boinovski <todorb@qti.qualcomm.com>

max-krasnyansky committed:
model: add support for extra bufs for all devices

max-krasnyansky committed:
hexagon: fix format checker errors

max-krasnyansky committed:
hexagon: add simple graph optimizer for stacking MUL_MAT ops with the same input

max-krasnyansky committed:
hexagon: update readme and cmake presets

max-krasnyansky committed:
ci: add android-ndk-build jobs that build plain ARM64 and Snapdragon versions

max-krasnyansky commented:
@slaren @bandoti @CISC 
Basic CI is in. I added `android-ndk-build` that builds vanilla ARM64 CPU and Snapdragon (CPU/GPU/NPU) flavors.
Those builds can be just pushed via ADB to the devices and run.
I'm going to hook it up to the [Qualcomm Device Cloud](https://qdc.qualcomm.com/) so that we can run jobs on the actual Snapdragon-based devices, but that will need some more work. I'm going to prototype in the separate repo first.

btw That `build.yml` is getting long. I kind of like how you guys did `linux-cross` builds in the separate yaml, it's collapsible in the UI, etc.
Should we do the same for android, ubuntu, windows, etc?

max-krasnyansky committed:
hexagon: move ADB helper scripts into scripts/snapdragon/adb

max-krasnyansky commented:
> > You can try to reuse the `ggml_graph_optimize` from the Metal backend:
 
@ggerganov  @jeffbolznv 
I added a simple optimizer to stack up the MUL_MATs with the same input and it's working great.
It ended up being pretty simple especially with the reuse from Metal backend. Thanks for the suggestions!

Quick question for you guys. 
Is the optimizer allowed to modify the nodes themselves? i.e would the mods be sticky and visible in `graph_compute`?
During the compute I need to know a couple of things. 1) wether the src1 can reused 2) what is the last compute op in the graph.
I'm doing both dynamically but the optimizer could just add some backend specific flags since it already knows all that.

max-krasnyansky committed:
hexagon: replace all f/printfs with GGML_LOG_...

max-krasnyansky committed:
readme: add hexagon to the list supported backends

jeffbolznv commented:
IMO the nodes should be treated as const.

ggerganov commented on ggml/src/ggml-hexagon/ggml-hexagon.cpp:
```diff
@@ -3211,6 +3211,167 @@ static void ggml_backend_hexagon_synchronize(ggml_backend_t backend) {
+    }
+
+    return res;
+}
```
I think this logic will incorrectly reorder the following matrix multiplications:

```
c = ggml_mul_mat(a, b);
d = ggml_scale(c);
e = ggml_mul_mat(d, b);
```

IMO, it's better to reuse the full logic for graph optimization through the `ggml_mem_ranges`. We can do this separately in a follow-up PR to not get distracted from the main purpose here.

max-krasnyansky replied:
Ah. Indeed it would reorder that sequence incorrectly. I convinced myself that I don't need to track the other inputs for the matmuls after looking at the graphs for the models I typically test with (listed in PR) but it's indeed not generic enough for all cases. I now added a check for src0 to be quantized, that should make it a bit safer and we do want to only stack those matmuls anyways. Ideally, it'd be good to have `ggml_is_weight()` in this case, which would make it totally safe, but we don't seem to have that anymore (I seem to recall seeing that somewhere). 

We might need more complicated fusion/stacking rules in general. For example I could fuse/stack these guys:

```
ggml-hex: HTP0 device-supports-op MUL  : norm-16 x blk.16.attn_norm.weight -> attn_norm-16 : ...
ggml-hex: HTP0 device-supports-op MUL_MAT : blk.16.attn_q.weight x attn_norm-16 -> Qcur-16 : ...
ggml-hex: HTP0 device-supports-op MUL_MAT : blk.16.attn_k.weight x attn_norm-16 -> Kcur-16 : ...
ggml-hex: HTP0 device-supports-op MUL_MAT : blk.16.attn_v.weight x attn_norm-16 -> Vcur-16 : ...
```
i.e Just have the MUL quantize its output straight into VTCM.
(that's from llama-3.2-3B)

jeffbolznv replied:
FWIW, the Vulkan backend reuses converted data across multiple nodes, e.g. converting the B matrix for MUL_MAT to fp16 or q8_1 and reusing it for the next few multiplies. This is handled entirely in the backend during graph_compute.

max-krasnyansky replied:
@jeffbolznv I'll take a look at how Vulkan backend does it but my guess right now that it's probably a bit different.
In the Hexagon case subsequent Ops might write other stuff into VTCM. Eventually I'll add more advanced VTCM manager/allocator, but for now each Op assumes it can use any part of it. Hence the need to re-order the MUL_MATs to simplify the reuse/tracking.

@ggerganov I just meant that reordering and fusion are related where one affects the other. 
i.e. If I stack/reorder the MUL_MATs like the above I can now essentially fuse those 4 ops. And yes, of course, I agree that to do that properly full dependencies of srcs/dsts need to be validated. I didn't fully grok your mem_ranges logic, sounds like it does exactly that. I'll stare at it again :). In the meantime I'm going to add the TODO.

max-krasnyansky committed:
hexagon: stack malmuts with quantized inputs only

max-krasnyansky committed:
hexagon: add TODO for fixing issues in hexagon_graph_optimize

max-krasnyansky committed:
hexagon: update to hex-sdk 6.4.0 and add scripts for running on QDC

max-krasnyansky commented:
@slaren Let me know if you had any more thoughts on the REPACK buffers.
Currently the main features are fully functional with that change in `llama-model` to allow for `extra_bufs` for all devs.
Everything works as expected with CPU/OpenCL/Hexagon and I checked my CPU/Metal and CPU/CUDA setups and don't see any obvious issues. 
One thing that I'd like to add soon-ish is the ability to override the tensors with HTP-REPACK buffers. I'm thinking for now I could just add a simple change (similar to `llama-model`) to allow for `extra_bufs` in that path. 
Basically, these changes are just allowing any backend to have `extra_bufs` and not just the CPU.

We could revisit things and come up with a better solution in the followup. I keep thinking it be nice if we could just keep everything as a host buffer and somehow enforce `set/get_tensor` for specific tensors. It's sort of what the non-host buffers do but non-host is kind of an overkill in the Hexagon case. ie It ends up being a separate memory mapping, etc just to force the `set/get_tensor` and REPACK. 

---
I made good progress on enabling Qualcomm Device Cloud (aka QDC) where we can run CI jobs on physical devices.
It does need a bit more work and setup. I'll reach out to you guys via email to setup some hidden CI vars using a private API_TOKEN.

max-krasnyansky committed:
scripts: fix lint errors

max-krasnyansky committed:
scripts: update qdc pytest script to make linter happy

max-krasnyansky committed:
hexagon: add reduce sum in fp32

max-krasnyansky committed:
hexagon: reduce number of vector stores in matmul output

max-krasnyansky committed:
hexagon: remove the need for vdelta in reduce-multiply-x8

max-krasnyansky committed:
hexagon: consistent use of reduce_sum_fp32 for row_sums

max-krasnyansky committed:
hexagon: some more matmul optimizations and comments

Optimize cases where tensor dims are not multiple of 1024 (e.g in Qwen models).
We've handled those cases already but at a higher overhead.

slaren commented on ggml/include/ggml-hexagon.h:
```diff
@@ -0,0 +1,19 @@
+
+GGML_API bool ggml_backend_is_hexagon(ggml_backend_t backend);
+
+GGML_API ggml_backend_reg_t ggml_backend_hexagon_reg(void);
```
These functions should use `GGML_BACKEND_API` instead.

slaren commented on ggml/src/ggml-hexagon/ggml-hexagon.cpp:
```diff
@@ -0,0 +1,3713 @@
+                                                   ggml_tensor *         tensor,
+                                                   const void *          data,
+                                                   size_t                offset,
+                                                   size_t                size) {
```
The repack functions seem to ignore the size, and assume that this function is called with the data of the full tensor. While this is true at the moment, it is a dangerous assumption to make and it should have at least an assert.

slaren commented on ggml/src/ggml-hexagon/ggml-hexagon.cpp:
```diff
@@ -0,0 +1,3713 @@
+    return true;
+
+    GGML_UNUSED(buffer);
+}
```
There is no guarantee that the `src` tensor is in a compatible buffer type, you need to check the buffer type and only perform the copy if it is from a known and supported buffer type.

max-krasnyansky commented:
@slaren sorry for bugging you.  Need your blessing here :) Should be ready to merge as the first cut.

slaren commented on ggml/src/ggml-hexagon/ggml-hexagon.cpp:
```diff
@@ -0,0 +1,3713 @@
+static void ggml_backend_hexagon_free(ggml_backend_t backend) {
+    auto sess = static_cast<ggml_hexagon_session *>(backend->context);
+    GGML_UNUSED(sess);
+}
```
Nothing in freed here, will cause a leak.

max-krasnyansky replied:
Yep. Another good catch. Fixed,

max-krasnyansky committed:
hexagon: update cmake presets

max-krasnyansky committed:
hexagon: add OPMASK support for run-bench.sh wrapper

max-krasnyansky committed:
hexagon: update to use GGML_BACKEND_API

max-krasnyansky committed:
hexagon: remove unused logic for setting tensor flags for the views

max-krasnyansky committed:
hexagon: add asserts to set/get_tensor to make sure we handle complete tensors

Same asserts as the CPU backend.

max-krasnyansky committed:
hexagon: use cpy_tensor slow path for non-host buffers

max-krasnyansky committed:
hexagon: error checks in the buffer allocator

max-krasnyansky committed:
cmake: move include(extProj) under ggml-hexagon

max-krasnyansky committed:
hexagon: don't forget to delete the backend on free

max-krasnyansky committed:
hexagon: set/get_tensor size assert apply only to quantized tensors

max-krasnyansky committed:
hexagon: reintroduce HEX_VERBOSE wrapper for GGML_LOG_DEBUG for now

GGML_LOG_DEBUG is always enabled for test-backend-ops and the output gets in the way.
Ideally we need a bit more finer log levels.

max-krasnyansky commented on ggml/src/ggml-hexagon/ggml-hexagon.cpp:
```diff
@@ -0,0 +1,3713 @@
+        if (!ctx->mapped) {
+            ctx->mmap();
+        }
+    }
```
Yep. My bad. Cut and paste from another backend. No need for flags and extra for now.

max-krasnyansky commented:
> > @slaren Let me know if you had any more thoughts on the REPACK buffers.
> > Currently the main features are fully functional with that change in `llama-model` to allow for `extra_bufs` for all devs.
> 
> This is good for now. It's just not going to be easy for other ggml applications to use this backend, since the logic to use extra buffer types is rather complex and entirely the responsibility of applications at the moment. We should improve that, but a fairly simple change that should allow doing repack in the backend default buffer type, and still avoid copies from/to the CPU backend for non-repacked types, would be to make the `is_host` function take a type parameter, rather than blanket applying it to the entire buffer type.

Actually the default use-case requires no special handling from the apps. Very much the same as OpenCL or any other backends.

```
$ adb shell
pa2q:/ $ cd /data/local/tmp/llama.cpp/
pa2q:/data/local/tmp/llama.cpp $ ls
pa2q:/data/local/tmp/llama.cpp $ export LD_LIBRARY_PATH=$(pwd)/lib
pa2q:/data/local/tmp/llama.cpp $ export ADSP_LIBRARY_PATH=$(pwd)/lib
pa2q:/data/local/tmp/llama.cpp $ ./bin/llama-cli -m ../gguf/Llama-3.2-1B-Instruct-Q4_0.gguf -t 4 -f ../some-prompt.txt -no-cnv
```

That's it. The repack buffers are used automatically, etc. Partial offloads (ngl), etc all that works as expected.
See full trace below. 

If both OpenCL and Hexagon backends are enabled then the user/app would have to explicitly specify the `--device` option to pick. ie Kind of like having two GPUs, Accelerators, etc.

But as I mentioned above I'd love to get rid of the REPACK buffer type here :) and I really like your suggestion for `is_host()`.
Though I'm thinking it might be better to pass the tensor pointer to it and not just the type. That way we could maybe future-proof it a bit. ie For now the backends will just use the tensor type but later we could add some other conditions (like maybe some tensors can share the layout). btw This will be useful for the OpenCL backend as well, all the buffers there are from unified memory as well. 
I'm going to prototype this and tag you for another PR.  

Thanks for the thorough (as always) review and suggestions! 

<details>

```
ggml-hex: Hexagon backend (experimental) : allocating new registry : ndev 1
ggml-hex: Hexagon Arch version v79
ggml-hex: allocating new session: HTP0
ggml-hex: new session: HTP0 : session-id 0 domain-id 3 uri file:///libggml-htp-v79.so?htp_iface_skel_handle_invoke&_modver=1.0&_dom=cdsp&_session=0 handle 0xb4000070ed647cd0
build: 6825 (f7acd542d) with Android (13324770, +pgo, +bolt, +lto, +mlgo, based on r530567d) clang version 19.0.0 (https://android.googlesource.com/toolchain/llvm-project 97a699bf4812a18fb657c2779f5296a4ab2694d2) for x86_64-unknown-linux-gnu
main: llama backend init
main: load the model and apply lora adapter, if any
...
oad_tensors: offloading 16 repeating layers to GPU
load_tensors: offloading output layer to GPU
load_tensors: offloaded 17/17 layers to GPU
load_tensors:   CPU_Mapped model buffer size =   249.14 MiB
load_tensors:         HTP0 model buffer size =     0.26 MiB
load_tensors:  HTP0-REPACK model buffer size =   504.00 MiB.     <<< THIS was selected automatically
...
llama_context: n_ctx_per_seq (4096) < n_ctx_train (131072) -- the full capacity of the model will not be utilized
llama_context:        CPU  output buffer size =     0.49 MiB
llama_kv_cache:       HTP0 KV buffer size =   128.00 MiB
llama_kv_cache: size =  128.00 MiB (  4096 cells,  16 layers,  1/1 seqs), K (f16):   64.00 MiB, V (f16):   64.00 MiB
llama_context: layer 0 is assigned to device HTP0 but the Flash Attention tensor is assigned to device CPU (usually due to missing support)
llama_context: Flash Attention was auto, set to disabled
llama_context:       HTP0 compute buffer size =   524.00 MiB
llama_context:        CPU compute buffer size =   268.01 MiB
llama_context: graph nodes  = 582
llama_context: graph splits = 73
...
llama_perf_sampler_print:    sampling time =       5.27 ms /   300 runs   (    0.02 ms per token, 56958.42 tokens per second)
llama_perf_context_print:        load time =     563.48 ms
llama_perf_context_print: prompt eval time =    1348.02 ms /   205 tokens (    6.58 ms per token,   152.07 tokens per second)
llama_perf_context_print:        eval time =    2014.48 ms /    94 runs   (   21.43 ms per token,    46.66 tokens per second)
llama_perf_context_print:       total time =    3484.24 ms /   299 tokens
llama_perf_context_print:    graphs reused =         93
llama_memory_breakdown_print: | memory breakdown [MiB] | total   free    self   model   context   compute    unaccounted |
```

</details>

max-krasnyansky commented:
> > > @slaren Let me know if you had any more thoughts on the REPACK buffers.
> > > Currently the main features are fully functional with that change in `llama-model` to allow for `extra_bufs` for all devs.
> > 
> > 
> > This is good for now. It's just not going to be easy for other ggml applications to use this backend, since the logic to use extra buffer types is rather complex and entirely the responsibility of applications at the moment. We should improve that, but a fairly simple change that should allow doing repack in the backend default buffer type, and still avoid copies from/to the CPU backend for non-repacked types, would be to make the `is_host` function take a type parameter, rather than blanket applying it to the entire buffer type.
> 
> Actually the default use-case requires no special handling from the apps. Very much the same as OpenCL or any other backends.

Re-read your reply again and you said _ggml applications_. Yeah, that's true. The app would have to query extra buffs explicitly. Definitely make sense to improve that via `is_host(tensor or type)` API.

slaren commented on ggml/src/ggml-hexagon/ggml-hexagon.cpp:
```diff
@@ -0,0 +1,3724 @@
+    } catch (std::exception const &exc) {
+        GGML_LOG_ERROR("ggml-hex: %s failed to allocate buffer context: %s\n", sess->name.c_str(), exc.what());
+        return nullptr;
+    }
```
I don't think this is enough, the `ggml_backend_hexagon_buffer_context` constructor does not throw when the buffer allocation fails.

ggerganov commented on ggml/src/ggml-hexagon/ggml-hexagon.cpp:
```diff
@@ -0,0 +1,3724 @@
+    auto now = std::chrono::high_resolution_clock::now().time_since_epoch();
+    return std::chrono::duration_cast<std::chrono::microseconds>(now).count();
+}
+
```
This is unused

max-krasnyansky replied:
Removed.

ggerganov commented on ggml/src/ggml-hexagon/ggml-hexagon.cpp:
```diff
@@ -0,0 +1,3724 @@
+    buffer_type.context = new ggml_backend_hexagon_buffer_type_context(this->name, this);
+
+    repack_buffer_type.iface   = ggml_backend_hexagon_repack_buffer_type_interface;
+    repack_buffer_type.context = new ggml_backend_hexagon_buffer_type_context(this->name + "-REPACK", this);
```
Are these allocations released?

max-krasnyansky committed:
docs: typos in hexagon developer docs (libggm-...)

max-krasnyansky committed:
hexagon: overhaul error handling in the session/device allocation

this should handle all failure paths in the session allocation.

max-krasnyansky committed:
hexagon: update cmake presets to enable fp16 vectors

max-krasnyansky committed:
hexagon: remove unused time_usec function

slaren commented on ggml/src/ggml-hexagon/htp/hvx-utils.c:
```diff
@@ -0,0 +1,944 @@
+                     const int num_elems) {
+    htp_binary_ops_preamble
+
+        for (int i = 0; i < step_of_4; i++) {
```
This looks unaligned.

```suggestion
    for (int i = 0; i < step_of_4; i++) {
```

max-krasnyansky replied:
Opps. I unleashed `clang-format` on the whole thing and missed that it got confused by missing semicolons after those preamble macros. Fixed.

max-krasnyansky committed:
hexagon: don't forget to release buffer contexts

max-krasnyansky committed:
hexagon: fixed indents in hvx-utils (missed clang-format auto-format failure)

max-krasnyansky committed:
hexagon: remove custom can_repeat function and use ggml_can_repeat
